For my project, I will be locating a possible location for a breakfast spot. I will be exploring hours of operations, proximity to other places that may increase traffic (e.g. churches, hotels), and density of other, competing restaurants. 

In [1]:
import numpy as np
import pandas as pd

import requests # library to handle requests
from pandas import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans

import folium

The first thing I'll need to do is define the location. I've chosen a midwestern U.S. city, Indianapolis, as the place I will be exploring, so I will need to find information about latitude and longitude. I found a website that will give coordinates for Zip Codes (https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/table/), so I narrowed it down to Indiana and exported a .csv file. 

In [2]:
inZipRaw = pd.read_csv("us-zip-code-latitude-and-longitude.csv", sep=";")
# We only need Indianapolis 
inCoord = inZipRaw[inZipRaw.City=="Indianapolis"].reset_index(drop=True)
print(inCoord['Zip'].count())
inCoord.head()

60


Zip          City State   Latitude  Longitude  Timezone  \
0  46227  Indianapolis    IN  39.678495  -86.12973        -5   
1  46201  Indianapolis    IN  39.775125  -86.10839        -5   
2  46250  Indianapolis    IN  39.905689  -86.06733        -5   
3  46228  Indianapolis    IN  39.849474  -86.20448        -5   
4  46224  Indianapolis    IN  39.795593  -86.25409        -5   

   Daylight savings time flag             geopoint  
0                           0  39.678495,-86.12973  
1                           0  39.775125,-86.10839  
2                           0  39.905689,-86.06733  
3                           0  39.849474,-86.20448  
4                           0  39.795593,-86.25409

Let's visualize these coordinates in the city of Indianapolis.

In [3]:
cityLat = 39.7684
cityLong = -86.1581
map_Ind = folium.Map(location=[cityLat,cityLong], zoom_start=11)
for lat, lng, zipcode in zip(inCoord['Latitude'], inCoord['Longitude'], inCoord['Zip']):
    label = '{}'.format(zipcode)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Ind)
map_Ind

Indianapolis is not as big as cities like Chicago or New York, so this distribution should be fine for our purposes. However, it is more spread out, so we'll need to have a wider radius. Next, we'll connect to FourSquare.

In [4]:
#FourSquare call
CLIENT_ID = '2DDTSCAMVOKHLPZTYIJSVR4TIVEXPIWIW0DM141PB5AXLMCP' 
CLIENT_SECRET = '1GYDYKTOJQS33PND3MEVRBV2N1FMNDLDDG13TB3BMJK3HW3M' 
VERSION = '20180605' 

## Venue Categories
We need to get both the type of venue and the popular operating hours. Because popular hours are a premium call and we can only do 500 a day, we'll start by narrowing down the neighborhoods by clustering via the venue categories. 
Let's start first with a test of a single ZIP.

In [5]:
first_zip_lat = inCoord.loc[0, 'Latitude'] 
first_zip_long = inCoord.loc[0, 'Longitude']
first_zip_name = inCoord.loc[0, 'Zip'] 

LIMIT = 50
radius = 2500 #2500 meters is approx 1.5 miles

# url for regular call
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    first_zip_lat, 
    first_zip_long, 
    radius, 
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ef4c4a0e9e7ad1512a3e184'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Indianapolis',
  'headerFullLocation': 'Indianapolis',
  'headerLocationGranularity': 'city',
  'totalResults': 74,
  'suggestedBounds': {'ne': {'lat': 39.700995022500024,
    'lng': -86.10055007922678},
   'sw': {'lat': 39.65599497749997, 'lng': -86.15890992077324}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4be2fe23d27a20a17fed905b',
       'name': 'CVS pharmacy',
       'location': {'address': '5920 Madison Ave',
        'lat': 39.680284044828916,
        'lng': -86.13270748942533,
        'labeledLatLngs': [{'label': 'display',
          'lat': 39.6802840448289

In [6]:
venues = results['response']['groups'][0]['items']
venues_normal = json_normalize(venues)
print(venues_normal['venue.id'].count())
venues_normal.head()

50


referralId  reasons.count  \
0  e-0-4be2fe23d27a20a17fed905b-0              0   
1  e-0-4b3b7406f964a5200f7425e3-1              0   
2  e-0-4bbf43eeba9776b00bc9fec8-2              0   
3  e-0-4b3e97bef964a520709f25e3-3              0   
4  e-0-50e32cbbe4b092e9baf01df1-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id                   venue.name  \
0  4be2fe23d27a20a17fed905b                 CVS pharmacy   
1  4b3b7406f964a5200f7425e3                Long's Bakery   
2  4bbf43eeba9776b00bc9fec8                       SUBWAY   
3  4b3e97bef964a520709f25e3                       Kroger   
4  50e32cbbe4b092e9baf01df1  Kim's Kakery, Bakery & Cafe   

   venue.location.address  venue.location.lat  venue.location.lng  \
0        5920 Madison Ave           39.680284          -86.132707   
1     2301 E Southport Rd           39.664902          -86.121751   
2  6025 Madison Ave Ste A           39.679301          -86.130128   
3        5911 Madison Ave           39.680406          -86.129356   
4        5452 Madison Ave           39.687068          -86.135175   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': 39.68028404482891...                      323   
1  [{'label': 'display', 'lat': 39.66490225215457...                     1660   
2  [{'label': 'display', 'lat': 39.67930094575346...                       95   
3  [{'label': 'display', 'lat': 39.6804065, 'lng'...                      215   
4  [{'label': 'display', 'lat': 39.68706779787860...                     1062   

   ...                                   venue.categories venue.photos.count  \
0  ...  [{'id': '4bf58dd8d48988d10f951735', 'name': 'P...                  0   
1  ...  [{'id': '4bf58dd8d48988d148941735', 'name': 'D...                  0   
2  ...  [{'id': '4bf58dd8d48988d1c5941735', 'name': 'S...                  0   
3  ...  [{'id': '52f2ab2ebcbc57f1066b8b46', 'name': 'S...                  0   
4  ...  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...                  0   

  venue.photos.groups venue.location.crossStreet venue.delivery.id  \
0                  []                        NaN               NaN   
1                  []                        NaN               NaN   
2                  []            E Edgewood Ave.           1109838   
3                  []           at E Edgewood Av               NaN   
4                  []                  Epler Ave           2103615   

                                  venue.delivery.url  \
0                                                NaN   
1                                                NaN   
2  https://www.grubhub.com/restaurant/subway-6025...   
3                                                NaN   
4  https://www.grubhub.com/restaurant/kims-kakery...   

  venue.delivery.provider.name       venue.delivery.provider.icon.prefix  \
0                          NaN                                       NaN   
1                          NaN                                       NaN   
2                      grubhub  https://fastly.4sqi.net/img/general/cap/   
3                          NaN                                       NaN   
4                      grubhub  https://fastly.4sqi.net/img/general/cap/   

  venue.delivery.provider.icon.sizes        venue.delivery.provider.icon.name  
0                                NaN                                      NaN  
1                                NaN                                      NaN  
2                           [40, 50]  /delivery_provider_grubhub_20180129.png  
3                                NaN                                      NaN  
4             

Let's use the get_category_type function as in the other assignments in order to extract the categories.

In [7]:
# function to get the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [8]:
# filter columns
filtered_columns = ['venue.name', 'venue.id', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
venues_normal = venues_normal.loc[:, filtered_columns]

# filter the category for each row
venues_normal['venue.categories'] = venues_normal.apply(get_category_type, axis=1)

# clean columns
venues_normal.columns = [col.split(".")[-1] for col in venues_normal.columns]

print('There are {} venues in ZIP {}.'.format(venues_normal.shape[0], first_zip_name))
venues_normal.head()

There are 50 venues in ZIP 46227.


name                        id      categories  \
0                 CVS pharmacy  4be2fe23d27a20a17fed905b        Pharmacy   
1                Long's Bakery  4b3b7406f964a5200f7425e3      Donut Shop   
2                       SUBWAY  4bbf43eeba9776b00bc9fec8  Sandwich Place   
3                       Kroger  4b3e97bef964a520709f25e3     Supermarket   
4  Kim's Kakery, Bakery & Cafe  50e32cbbe4b092e9baf01df1          Bakery   

         lat        lng  
0  39.680284 -86.132707  
1  39.664902 -86.121751  
2  39.679301 -86.130128  
3  39.680406 -86.129356  
4  39.687068 -86.135175

Everything's looking good--let's repeat this for all of the venues. We'll use the getNearbyVenues function as in previous assignments.

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=2500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'],
            v['venue']['id'],
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue',
                  'Venue ID',                             
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
indy_venues = getNearbyVenues(names=inCoord['Zip'],
                                   latitudes=inCoord['Latitude'],
                                   longitudes=inCoord['Longitude']
                                  )

46227
46201
46250
46228
46224
46216
46242
46244
46231
46259
46295
46234
46221
46290
46249
46222
46235
46241
46203
46236
46204
46209
46278
46251
46230
46217
46214
46285
46256
46223
46280
46275
46218
46274
46205
46291
46255
46202
46283
46208
46211
46219
46207
46254
46229
46266
46247
46220
46253
46237
46240
46298
46239
46260
46277
46225
46282
46268
46226
46206


In [18]:
print(indy_venues.shape)
display(indy_venues.head())
indy_venues.groupby('Neighborhood').count()

(2758, 8)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         46227              39.678495               -86.12973   
1         46227              39.678495               -86.12973   
2         46227              39.678495               -86.12973   
3         46227              39.678495               -86.12973   
4         46227              39.678495               -86.12973   

                         Venue                  Venue ID  Venue Latitude  \
0                 CVS pharmacy  4be2fe23d27a20a17fed905b       39.680284   
1                Long's Bakery  4b3b7406f964a5200f7425e3       39.664902   
2                       SUBWAY  4bbf43eeba9776b00bc9fec8       39.679301   
3                       Kroger  4b3e97bef964a520709f25e3       39.680406   
4  Kim's Kakery, Bakery & Cafe  50e32cbbe4b092e9baf01df1       39.687068   

   Venue Longitude  Venue Category  
0       -86.132707        Pharmacy  
1       -86.121751      Donut Shop  
2       -86.130128  Sandwich Place  
3       -86.129356     Supermarket  
4       -86.135175          Bakery

Neighborhood Latitude  Neighborhood Longitude  Venue  Venue ID  \
Neighborhood                                                                   
46201                            50                      50     50        50   
46202                            50                      50     50        50   
46203                            50                      50     50        50   
46204                            50                      50     50        50   
46205                            50                      50     50        50   
46206                            50                      50     50        50   
46207                            50                      50     50        50   
46208                            47                      47     47        47   
46209                            50                      50     50        50   
46211                            50                      50     50        50   
46214                            50                      50     50        50   
46216                            50                      50     50        50   
46217                            42                      42     42        42   
46218                            30                      30     30        30   
46219                            50                      50     50        50   
46220                            50                      50     50        50   
46221                            37                      37     37        37   
46222                            50                      50     50        50   
46223                            50                      50     50        50   
46224                            50                      50     50        50   
46225                            50                      50     50        50   
46226                            50                      50     50        50   
46227                            50                      50     50        50   
46228                            24                      24     24        24   
46229                            49                      49     49        49   
46230                            50                      50     50        50   
46231                            50                      50     50        50   
46234                            33                      33     33        33   
46235                            29                      29     29        29   
46236                            50                      50     50        50   
46237                            50                      50     50        50   
46239                            13                      13     13        13   
46240                            50                      50     50        50   
46241                            50                      50     50        50   
46242                            50                      50     50        50   
46244                            50                      50     50        50   
46247                            50                      50     50        50   
46249                            49                      49     49        49   
46250                            50                      50     50        50   
46251                            50                      50     50        50   
46253                            50                      50     50        50   
46254                            50                      50     50        50   
46255                            50                      50     50        50   
46256                            50                      50     50        50   
46259                             8                       8      8         8   
46260                            32                      32     32        32   
46266                            50                      50     50        50   
46268                            50                      50     50        50   
46274         

In [19]:
print('There are {} uniques categories.'.format(len(indy_venues['Venue Category'].unique())))

There are 240 uniques categories.


There are many unique categories, but in reality, many of these won't matter to us when we're looking to open a breakfast spot. We'll take a look at the venues and narrow down a list.  

In [20]:
venue_type_all = indy_venues['Venue Category'].unique()
venue_type_all

array(['Pharmacy', 'Donut Shop', 'Sandwich Place', 'Supermarket',
       'Bakery', 'Chinese Restaurant', 'Coffee Shop', 'Gas Station',
       'Breakfast Spot', 'Antique Shop', 'Salon / Barbershop',
       'Seafood Restaurant', 'Lingerie Store', 'American Restaurant',
       'Mexican Restaurant', 'Bar', 'Indian Restaurant', 'Pizza Place',
       'Scenic Lookout', 'Greek Restaurant', 'Discount Store',
       'Video Game Store', 'Fast Food Restaurant', 'Bank',
       'Thai Restaurant', 'Pub', 'Video Store', 'Cosmetics Shop',
       'Ice Cream Shop', 'Big Box Store', 'Kids Store', 'Dance Studio',
       'Gym', 'Grocery Store', 'Fried Chicken Joint', 'Park', 'Café',
       'Diner', 'Liquor Store', 'Department Store',
       'Gym / Fitness Center', 'Shopping Mall', 'Food Truck',
       'Flea Market', 'Gay Bar', 'Trail', 'Vietnamese Restaurant',
       'Sporting Goods Shop', 'Clothing Store', 'Comic Shop',
       'Sports Bar', 'Fondue Restaurant', 'Health Food Store',
       'Record Shop', 'T

From this list, I manually narrowed down the type of venues that seem most likely to either attract people during the daytime or morning (such as tourist attractions but excluding late night activities like bars) or be in competition with a breakfast spot (excluding dinner restaurants which likely wouldn't attract morning crowds or detract from those crowds). While this step could be done in a more data-drive capacity, the limited number of premium calls means it would be better to narrow down our areas earlier on in such a way that we can choose which zip codes to pursue for venue hours more easily.

In [21]:
venue_type_filter = ['Donut Shop', 'Sandwich Place','Bakery', 'Coffee Shop', 'Gas Station', 'Breakfast Spot', 'Antique Shop', 'American Restaurant', 'Mexican Restaurant', 'Fast Food Restaurant', 'Big Box Store', 'Park', 'Café', 'Diner', 'Department Store', 'Shopping Mall', 'Food Truck', 'Trail', 'Health Food Store', 'Deli / Bodega', 'Fabric Shop', 'Cafeteria', 'Golf Course', 'Convenience Store', 'Food', 'Racetrack',  'Museum', 'Playground', 'Farmers Market', 'Monument / Landmark', 'Campground', 'Hotel', 'State / Provincial Park', 'Soccer Stadium',  'Government Building', 'Bike Trail', 'Bookstore', 'Arts & Crafts Store', 'Juice Bar', 'New American Restaurant', 'Concert Hall', 'Southern / Soul Food Restaurant', 'Restaurant', 'Supplement Shop', "Women's Store", 'Optical Shop', 'Business Service', 'Other Repair Shop', 'Construction & Landscaping', 'Flower Shop', 'Real Estate Office', 'School', 'Climbing Gym', 'Rental Car Location', 'Tennis Court', 'Locksmith', 'Dog Run', 'History Museum', 'Thrift / Vintage Store', 'Pawn Shop',  'Rental Service', 'Pet Service', 'Organic Grocery', 'Vegetarian / Vegan Restaurant', 'Fountain', 'Garden', 'Art Gallery', 'Art Museum', 'Outdoor Supply Store', 'Office', 'Post Office', 'Outdoors & Recreation', 'Plaza', 'Performing Arts Venue', 'Hardware Store', 'Basketball Stadium', 'Building', 'Shopping Plaza', 'Sculpture Garden', 'Baseball Field', 'Athletics & Sports', 'Volleyball Court', 'Skating Rink', 'Automotive Shop', 'Home Service', 'Hobby Shop', 'Truck Stop', 'Motel', 'Snack Place', 'Pool', 'Recreation Center', 'Kitchen Supply Store', 'Market', 'Hockey Arena', 'Fair', 'Farm', 'General Entertainment', 'Science Museum', 'Historic Site', 'Planetarium', 'Theme Park', 'Hostel', 'Other Great Outdoors', 'River', 'Bike Shop', 'Baseball Stadium', 'Football Stadium',  'Soccer Field', 'Basketball Court', 'Martial Arts Dojo', 'Multiplex', 'Bagel Shop', 'Food & Drink Shop', 'Outlet Store', 'Water Park', 'Boutique', 'Photography Studio',  'Food Court', 'Health & Beauty Service']
indy_venues_filtered = indy_venues[indy_venues["Venue Category"].isin(venue_type_filter)]

### Analyse and cluster
Now that we have our venues and the categories we need, let's cluster by ZIP code. 

In [22]:
# one hot encoding
indy_onehot = pd.get_dummies(indy_venues_filtered[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
indy_onehot['Neighborhood'] = indy_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [indy_onehot.columns[-1]] + list(indy_onehot.columns[:-1])
indy_onehot = indy_onehot[fixed_columns]

indy_onehot.head()

Neighborhood  American Restaurant  Antique Shop  Art Gallery  Art Museum  \
1         46227                    0             0            0           0   
2         46227                    0             0            0           0   
4         46227                    0             0            0           0   
6         46227                    0             0            0           0   
7         46227                    0             0            0           0   

   Arts & Crafts Store  Athletics & Sports  Automotive Shop  Bagel Shop  \
1                    0                   0                0           0   
2                    0                   0                0           0   
4                    0                   0                0           0   
6                    0                   0                0           0   
7                    0                   0                0           0   

   Bakery  ...  Supplement Shop  Tennis Court  Theme Park  \
1       0  ...                0             0           0   
2       0  ...                0             0           0   
4       1  ...                0             0           0   
6       0  ...                0             0           0   
7       0  ...                0             0           0   

   Thrift / Vintage Store  Trail  Truck Stop  Vegetarian / Vegan Restaurant  \
1                       0      0           0                              0   
2                       0      0           0                              0   
4                       0      0           0                              0   
6                       0      0           0                              0   
7                       0      0           0                              0   

   Volleyball Court  Water Park  Women's Store  
1                 0           0              0  
2                 0           0              0  
4                 0           0              0  
6                 0           0              0  
7                 0           0              0  

[5 rows x 120 columns]

In [23]:
indy_grouped = indy_onehot.groupby('Neighborhood').mean().reset_index()
indy_grouped

Neighborhood  American Restaurant  Antique Shop  Art Gallery  Art Museum  \
0          46201             0.000000      0.000000     0.000000    0.000000   
1          46202             0.050000      0.000000     0.000000    0.000000   
2          46203             0.000000      0.000000     0.105263    0.052632   
3          46204             0.043478      0.000000     0.000000    0.000000   
4          46205             0.125000      0.000000     0.000000    0.000000   
5          46206             0.074074      0.000000     0.000000    0.037037   
6          46207             0.035714      0.000000     0.000000    0.000000   
7          46208             0.034483      0.000000     0.034483    0.034483   
8          46209             0.100000      0.100000     0.000000    0.000000   
9          46211             0.100000      0.100000     0.000000    0.000000   
10         46214             0.041667      0.000000     0.000000    0.000000   
11         46216             0.074074      0.000000     0.000000    0.000000   
12         46217             0.055556      0.000000     0.000000    0.000000   
13         46218             0.000000      0.000000     0.000000    0.000000   
14         46219             0.166667      0.000000     0.000000    0.000000   
15         46220             0.047619      0.000000     0.000000    0.000000   
16         46221             0.000000      0.000000     0.000000    0.000000   
17         46222             0.000000      0.000000     0.000000    0.000000   
18         46223             0.100000      0.100000     0.000000    0.000000   
19         46224             0.060606      0.000000     0.000000    0.000000   
20         46225             0.040000      0.000000     0.080000    0.000000   
21         46226             0.045455      0.000000     0.000000    0.000000   
22         46227             0.066667      0.066667     0.000000    0.000000   
23         46228             0.083333      0.000000     0.000000    0.000000   
24         46229             0.100000      0.000000     0.000000    0.000000   
25         46230             0.100000      0.100000     0.000000    0.000000   
26         46231             0.000000      0.000000     0.000000    0.000000   
27         46234             0.000000      0.000000     0.000000    0.000000   
28         46235             0.125000      0.000000     0.000000    0.000000   
29         46236             0.105263      0.000000     0.000000    0.000000   
30         46237             0.200000      0.000000     0.000000    0.000000   
31         46239             0.000000      0.000000     0.000000    0.000000   
32         46240             0.050000      0.000000     0.000000    0.000000   
33         46241             0.064516      0.000000     0.000000    0.000000   
34         46242             0.100000      0.100000     0.000000    0.000000   
35         46244             0.100000      0.100000     0.000000    0.000000   
36         46247             0.100000      0.100000     0.000000    0.000000   
37         46249             0.136364      0.000000     0.000000    0.000000   
38         46250             0.083333      0.000000     0.000000    0.000000   
39         46251             0.100000      0.100000     0.000000    0.000000   
40         46253             0.100000      0.100000     0.000000    0.000000   
41         46254             0.117647      0.000000     0.000000    0.000000   
42         46255             0.100000      0.100000     0.000000    0.000000   
43         46256             0.045455      0.000000     0.000000    0.000000   
44         46259             0.000000      0.000000     0.000000    0.000000   
45         46260             0.142857      0.000000     0.000000    0.000000   
46         46266             0.100000      0.100000     0.000000    0.000000   
47         46268             0.000000      0.000000     0.000000    0.000000   
48         46274             0.100000      0.100000     0.000000    0.000000   
49  

Let's use the return_most_common_venues function from previous assignments to get the top venues for each ZIP code.

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [25]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = indy_grouped['Neighborhood']

for ind in np.arange(indy_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(indy_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0         46201  Fast Food Restaurant        Sandwich Place   
1         46202    Mexican Restaurant                 Plaza   
2         46203        Sandwich Place           Art Gallery   
3         46204    Mexican Restaurant                 Plaza   
4         46205   American Restaurant   Rental Car Location   

             3rd Most Common Venue    4th Most Common Venue  \
0                            Diner       Mexican Restaurant   
1  Southern / Soul Food Restaurant  New American Restaurant   
2                      Gas Station     Fast Food Restaurant   
3                      Coffee Shop                     Café   
4                   Science Museum           Sandwich Place   

           5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                           Park           Gas Station           Coffee Shop   
1                          Trail           Coffee Shop   Monument / Landmark   
2  Vegetarian / Vegan Restaurant            Art Museum                  Park   
3                 Breakfast Spot   American Restaurant    Basketball Stadium   
4             Athletics & Sports                  Café                  Fair   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                  Café       Department Store             Food Truck  
1         Deli / Bodega           Concert Hall                   Café  
2       Organic Grocery                   Café              Bookstore  
3        Hardware Store    Monument / Landmark                   Park  
4                  Farm  General Entertainment          Historic Site

Now it's time to cluster using k-means.

In [26]:
# set number of clusters
kclusters = 5

indy_grouped_clustering = indy_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(indy_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 1, 4, 2, 2, 2, 2, 2, 1, 1])

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted

Cluster Labels  Neighborhood       1st Most Common Venue  \
0                4         46201        Fast Food Restaurant   
1                1         46202          Mexican Restaurant   
2                4         46203              Sandwich Place   
3                2         46204          Mexican Restaurant   
4                2         46205         American Restaurant   
5                2         46206                       Hotel   
6                2         46207                       Hotel   
7                2         46208               Historic Site   
8                1         46209         American Restaurant   
9                1         46211         American Restaurant   
10               4         46214        Fast Food Restaurant   
11               4         46216              Breakfast Spot   
12               2         46217            Business Service   
13               3         46218                 Gas Station   
14               2         46219         American Restaurant   
15               2         46220                      Bakery   
16               3         46221        Fast Food Restaurant   
17               0         46222                   Racetrack   
18               1         46223         American Restaurant   
19               0         46224                   Racetrack   
20               2         46225          Mexican Restaurant   
21               3         46226        Fast Food Restaurant   
22               4         46227              Sandwich Place   
23               3         46228                 Golf Course   
24               4         46229                 Gas Station   
25               1         46230         American Restaurant   
26               4         46231            Department Store   
27               0         46234              Sandwich Place   
28               4         46235              Sandwich Place   
29               4         46236        Fast Food Restaurant   
30               4         46237         American Restaurant   
31               4         46239              Baseball Field   
32               2         46240              Sandwich Place   
33               4         46241                       Hotel   
34               1         46242         American Restaurant   
35               1         46244         American Restaurant   
36               1         46247         American Restaurant   
37               4         46249              Sandwich Place   
38               4         46250              Sandwich Place   
39               1         46251         American Restaurant   
40               1         46253         American Restaurant   
41               4         46254         American Restaurant   
42               1         46255         American Restaurant   
43               4         46256              Sandwich Place   
44               2         46259  Construction & Landscaping   
45               2         46260         American Restaurant   
46               1         46266         American Restaurant   
47               4         46268          Mexican Restaurant   
48               1         46274         American Restaurant   
49               1         46275         American Restaurant   
50               1         46277         American Restaurant   
51               2         46278                       Hotel   
52               2         46280                      Bakery   
53               1         46282         American Restaurant   
54               1         46283         American Restaurant   
55               1         46285         American Restaurant   
56               2         46290                 Coffee Shop   
57               1         46291         American Restaurant   
58               1         46295         American Restaurant   
59               1         46298         American Restaurant   

   2nd Most Common Venue            3rd Most Common Venue  \
0         Sandwich Place              

In [28]:
indy_merged = inCoord[['Zip', 'Latitude', 'Longitude']]
indy_merged = indy_merged.merge(neighborhoods_venues_sorted, left_on = "Zip", right_on = "Neighborhood")

Let's visualize the clusters on the map.

In [29]:
# create map
map_clusters = folium.Map(location=[cityLat, cityLong], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(indy_merged['Latitude'], indy_merged['Longitude'], indy_merged['Neighborhood'], indy_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's take a look at all of the clusters to get a sense of what's in them. 

In [30]:
indy_merged.loc[indy_merged['Cluster Labels'] == 0, indy_merged.columns[[1] + list(range(5, indy_merged.shape[1]))]]

Latitude 1st Most Common Venue 2nd Most Common Venue  \
4   39.795593             Racetrack    Mexican Restaurant   
11  39.797622        Sandwich Place           Gas Station   
15  39.786793             Racetrack           Golf Course   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4         Sandwich Place   American Restaurant      Department Store   
11             Racetrack           Coffee Shop           Golf Course   
15           Gas Station  Fast Food Restaurant                 Hotel   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4             Campground        Breakfast Spot           Golf Course   
11                School            Food Truck  Fast Food Restaurant   
15   Monument / Landmark        Soccer Stadium                  Park   

   9th Most Common Venue 10th Most Common Venue  
4    Monument / Landmark                 Museum  
11           Flower Shop     Real Estate Office  
15   Rental Car Location         Sandwich Place

In [31]:
indy_merged.loc[indy_merged['Cluster Labels'] == 1, indy_merged.columns[[1] + list(range(5, indy_merged.shape[1]))]]

Latitude 1st Most Common Venue 2nd Most Common Venue  \
6   39.779492   American Restaurant           Coffee Shop   
7   39.779492   American Restaurant           Coffee Shop   
10  39.779492   American Restaurant           Coffee Shop   
21  39.779492   American Restaurant           Coffee Shop   
23  39.779492   American Restaurant           Coffee Shop   
24  39.779492   American Restaurant           Coffee Shop   
27  39.779492   American Restaurant           Coffee Shop   
29  39.779492   American Restaurant           Coffee Shop   
31  39.779492   American Restaurant           Coffee Shop   
33  39.779492   American Restaurant           Coffee Shop   
35  39.779492   American Restaurant           Coffee Shop   
36  39.779492   American Restaurant           Coffee Shop   
37  39.782842    Mexican Restaurant                 Plaza   
38  39.779492   American Restaurant           Coffee Shop   
40  39.779492   American Restaurant           Coffee Shop   
45  39.779492   American Restaurant           Coffee Shop   
46  39.779492   American Restaurant           Coffee Shop   
48  39.779492   American Restaurant           Coffee Shop   
51  39.779492   American Restaurant           Coffee Shop   
54  39.779492   American Restaurant           Coffee Shop   
56  39.779492   American Restaurant           Coffee Shop   

              3rd Most Common Venue    4th Most Common Venue  \
6                              Café       Mexican Restaurant   
7                              Café       Mexican Restaurant   
10                             Café       Mexican Restaurant   
21                             Café       Mexican Restaurant   
23                             Café       Mexican Restaurant   
24                             Café       Mexican Restaurant   
27                             Café       Mexican Restaurant   
29                             Café       Mexican Restaurant   
31                             Café       Mexican Restaurant   
33                             Café       Mexican Restaurant   
35                             Café       Mexican Restaurant   
36                             Café       Mexican Restaurant   
37  Southern / Soul Food Restaurant  New American Restaurant   
38                             Café       Mexican Restaurant   
40                             Café       Mexican Restaurant   
45                             Café       Mexican Restaurant   
46                             Café       Mexican Restaurant   
48                             Café       Mexican Restaurant   
51                             Café       Mexican Restaurant   
54                             Café       Mexican Restaurant   
56                             Café       Mexican Restaurant   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6   New American Restaurant          Antique Shop                 Trail   
7   New American Restaurant          Antique Shop                 Trail   
10  New American Restaurant          Antique Shop                 Trail   
21  New American Restaurant          Antique Shop                 Trail   
23  New American Restaurant          Antique Shop                 Trail   
24  New American Restaurant          Antique Shop                 Trail   
27  New American Restaurant          Antique Shop                 Trail   
29  New American Restaurant          Antique Shop                 Trail   
31  New American Restaurant          Antique Shop                 Trail   
33  New American Restaurant          Antique Shop                 Trail   
35  New American Restaurant          Antique Shop                 Trail   
36  New American Restaurant          Antique Shop                 Trail   
37                    Trail           Coffee Shop   Monument / Landmark   
38  New American Restaurant          Antique Shop                 Trail   
40  New American Restaurant          Antique Shop                 Trail   
45  New American Restaurant          Antique Shop    

In [32]:
indy_merged.loc[indy_merged['Cluster Labels'] == 2, indy_merged.columns[[1] + list(range(5, indy_merged.shape[1]))]]

Latitude       1st Most Common Venue 2nd Most Common Venue  \
9   39.651145  Construction & Landscaping                 Diner   
13  39.934949                 Coffee Shop        Sandwich Place   
20  39.771743          Mexican Restaurant                 Plaza   
22  39.897920                       Hotel   American Restaurant   
25  39.668795            Business Service   American Restaurant   
30  39.939102                      Bakery   American Restaurant   
34  39.824858         American Restaurant   Rental Car Location   
39  39.820708               Historic Site           Golf Course   
41  39.780010         American Restaurant        Sandwich Place   
42  39.767293                       Hotel                 Plaza   
47  39.866913                      Bakery               Dog Run   
50  39.902039              Sandwich Place           Coffee Shop   
53  39.898107         American Restaurant           Coffee Shop   
55  39.746993          Mexican Restaurant           Art Gallery   
59  39.761293                       Hotel    Mexican Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
9                  Trail                  Park      Business Service   
13   American Restaurant           Gas Station     Convenience Store   
20           Coffee Shop                  Café        Breakfast Spot   
22            Food Truck        Shopping Plaza            Donut Shop   
25        Baseball Field          Home Service    Mexican Restaurant   
30   Rental Car Location        Breakfast Spot                 Hotel   
34        Science Museum        Sandwich Place    Athletics & Sports   
39           Planetarium                 Trail                  Park   
41           Coffee Shop                  Park   Rental Car Location   
42        History Museum    Mexican Restaurant                  Café   
47   American Restaurant                  Café        Breakfast Spot   
50        Breakfast Spot   American Restaurant    Mexican Restaurant   
53    Mexican Restaurant        Breakfast Spot           Golf Course   
55                 Hotel        Sandwich Place        Breakfast Spot   
59   American Restaurant        History Museum              Boutique   

    6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
9       Other Repair Shop     Food & Drink Shop                  Dog Run   
13    Rental Car Location           Flower Shop           Breakfast Spot   
20    American Restaurant    Basketball Stadium           Hardware Store   
22                   Park                  Café           Breakfast Spot   
25                Dog Run            Campground           Breakfast Spot   
30            Golf Course           Flower Shop        Other Repair Shop   
34                   Café                  Fair                     Farm   
39                  River        Science Museum         Sculpture Garden   
41         Breakfast Spot                 Diner               Restaurant   
42         Breakfast Spot   American Restaurant       Basketball Stadium   
47         Sandwich Place             Multiplex        Food & Drink Shop   
50       Department Store                  Café  New American Restaurant   
53         Sandwich Place                 Diner                Bookstore   
55    American Restaurant    Basketball Stadium            Deli / Bodega   
59  General Entertainment      Football Stadium                     Park   

    9th Most Common Venue 10th Most Common Venue  
9              Donut Shop            Fabric Shop  
13     Mexican Restaurant          Deli / Bodega  
20    Monument / Landmark                   Park  
22         Sandwich Place       Sculpture Garden  
25         Sandwich Place           Skating Rink  
30                   Park          Deli / Bodega  
34  General Entertainment          Historic Site  
39                 Museum                 Office  
41              Bike Shop         Farmers Market  
42  General Entertainment       Football Stadium  
4

In [33]:
indy_merged.loc[indy_merged['Cluster Labels'] == 3, indy_merged.columns[[1] + list(range(5, indy_merged.shape[1]))]]

Latitude 1st Most Common Venue 2nd Most Common Venue  \
3   39.849474           Golf Course           Gas Station   
12  39.719444  Fast Food Restaurant           Gas Station   
32  39.805841           Gas Station            Playground   
58  39.837290  Fast Food Restaurant           Gas Station   

   3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
3    American Restaurant     Convenience Store                 Coffee Shop   
12         Women's Store           Golf Course  Construction & Landscaping   
32                Market  Kitchen Supply Store        Fast Food Restaurant   
58   American Restaurant     Convenience Store                   Pawn Shop   

   6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
3                   Park  Fast Food Restaurant                    Food   
12          Climbing Gym        Sandwich Place      Mexican Restaurant   
32                  Park                  Food  Thrift / Vintage Store   
58                  Park            Donut Shop     Rental Car Location   

   9th Most Common Venue 10th Most Common Venue  
3      Food & Drink Shop                Dog Run  
12          Optical Shop                  Diner  
32                  Pool             Donut Shop  
58     Other Repair Shop         Sandwich Place

In [34]:
indy_merged.loc[indy_merged['Cluster Labels'] == 4, indy_merged.columns[[1] + list(range(5, indy_merged.shape[1]))]]

Latitude 1st Most Common Venue 2nd Most Common Venue  \
0   39.678495        Sandwich Place           Gas Station   
1   39.775125  Fast Food Restaurant        Sandwich Place   
2   39.905689        Sandwich Place      Department Store   
5   39.857639        Breakfast Spot        Sandwich Place   
8   39.719620      Department Store           Coffee Shop   
14  39.858989        Sandwich Place   American Restaurant   
16  39.835369        Sandwich Place   American Restaurant   
17  39.736844                 Hotel  Fast Food Restaurant   
18  39.742593        Sandwich Place           Art Gallery   
19  39.888225  Fast Food Restaurant           Golf Course   
26  39.792993  Fast Food Restaurant        Sandwich Place   
28  39.907838        Sandwich Place           Coffee Shop   
43  39.844024   American Restaurant           Coffee Shop   
44  39.795025           Gas Station        Sandwich Place   
49  39.678590   American Restaurant           Coffee Shop   
52  39.732943        Baseball Field        Sandwich Place   
57  39.897548    Mexican Restaurant        Sandwich Place   

   3rd Most Common Venue 4th Most Common Venue          5th Most Common Venue  \
0     Mexican Restaurant   American Restaurant                 Breakfast Spot   
1                  Diner    Mexican Restaurant                           Park   
2    American Restaurant    Mexican Restaurant                    Coffee Shop   
5   Fast Food Restaurant    Mexican Restaurant                         Bakery   
8   Fast Food Restaurant            Restaurant             Mexican Restaurant   
14  Fast Food Restaurant                Bakery                          Hotel   
16         Shopping Mall  Fast Food Restaurant                    Gas Station   
17        Sandwich Place   American Restaurant         Thrift / Vintage Store   
18           Gas Station  Fast Food Restaurant  Vegetarian / Vegan Restaurant   
19        Sandwich Place    Mexican Restaurant            American Restaurant   
26    Mexican Restaurant      Department Store                    Coffee Shop   
28    Mexican Restaurant       Supplement Shop                  Women's Store   
43    Mexican Restaurant        Sandwich Place                   Soccer Field   
44           Fabric Shop   American Restaurant                          Trail   
49        Sandwich Place                 Hotel             Mexican Restaurant   
52  Fast Food Restaurant           Golf Course                     Water Park   
57  Fast Food Restaurant       Supplement Shop                 Breakfast Spot   

              6th Most Common Venue   7th Most Common Venue  \
0                     Big Box Store    Fast Food Restaurant   
1                       Gas Station             Coffee Shop   
2                       Fabric Shop               Cafeteria   
5               American Restaurant          Soccer Stadium   
8                       Gas Station            Optical Shop   
14              Government Building             Golf Course   
16               Mexican Restaurant             Golf Course   
17                      Gas Station             Coffee Shop   
18                       Art Museum                    Park   
19                      Coffee Shop  Thrift / Vintage Store   
26  Southern / Soul Food Restaurant       Convenience Store   
28                      Snack Place                  Bakery   
43                          Dog Run     Rental Car Location   
44            Performing Arts Venue            Optical Shop   
49                      Gas Station            Outlet Store   
52                           School                    Park   
57                   Rental Service              Food Court   

   8th Most Common Venue    9th Most Common Venue   10th Most Common Venue  
0                 Bakery              Coffee Shop               Donut Shop  
1                   Café         Department Store               Food Truck  
2   Fast Food Restaurant            Deli / Bodega        Health Food Store  
5   

It looks like cluster label 2 has the best spread of businesses that include hotels and other promising businesses without too many competing coffee shops, cafes, bakeries, and so on. Let's take a closer look. 

In [35]:
top_zip = indy_merged.loc[indy_merged['Cluster Labels'] == 2].reset_index(drop=True)
top_zip

Zip   Latitude  Longitude  Cluster Labels  Neighborhood  \
0   46259  39.651145 -85.980730               2         46259   
1   46290  39.934949 -86.162620               2         46290   
2   46204  39.771743 -86.155980               2         46204   
3   46278  39.897920 -86.286190               2         46278   
4   46217  39.668795 -86.183300               2         46217   
5   46280  39.939102 -86.138310               2         46280   
6   46205  39.824858 -86.138170               2         46205   
7   46208  39.820708 -86.171300               2         46208   
8   46219  39.780010 -86.048890               2         46219   
9   46207  39.767293 -86.160616               2         46207   
10  46220  39.866913 -86.123230               2         46220   
11  46240  39.902039 -86.129480               2         46240   
12  46260  39.898107 -86.182320               2         46260   
13  46225  39.746993 -86.159030               2         46225   
14  46206  39.761293 -86.161336               2         46206   

         1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0   Construction & Landscaping                 Diner                 Trail   
1                  Coffee Shop        Sandwich Place   American Restaurant   
2           Mexican Restaurant                 Plaza           Coffee Shop   
3                        Hotel   American Restaurant            Food Truck   
4             Business Service   American Restaurant        Baseball Field   
5                       Bakery   American Restaurant   Rental Car Location   
6          American Restaurant   Rental Car Location        Science Museum   
7                Historic Site           Golf Course           Planetarium   
8          American Restaurant        Sandwich Place           Coffee Shop   
9                        Hotel                 Plaza        History Museum   
10                      Bakery               Dog Run   American Restaurant   
11              Sandwich Place           Coffee Shop        Breakfast Spot   
12         American Restaurant           Coffee Shop    Mexican Restaurant   
13          Mexican Restaurant           Art Gallery                 Hotel   
14                       Hotel    Mexican Restaurant   American Restaurant   

   4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
0                   Park      Business Service      Other Repair Shop   
1            Gas Station     Convenience Store    Rental Car Location   
2                   Café        Breakfast Spot    American Restaurant   
3         Shopping Plaza            Donut Shop                   Park   
4           Home Service    Mexican Restaurant                Dog Run   
5         Breakfast Spot                 Hotel            Golf Course   
6         Sandwich Place    Athletics & Sports                   Café   
7                  Trail                  Park                  River   
8                   Park   Rental Car Location         Breakfast Spot   
9     Mexican Restaurant                  Café         Breakfast Spot   
10                  Café        Breakfast Spot         Sandwich Place   
11   American Restaurant    Mexican Restaurant       Department Store   
12        Breakfast Spot           Golf Course         Sandwich Place   
13        Sandwich Place        Breakfast Spot    American Restaurant   
14        History Museum              Boutique  General Entertainment   

   7th Most Common Venue    8th Most Common Venue  9th Most Common Venue  \
0      Food & Drink Shop                  Dog Run             Donut Shop   
1            Flower Shop           Breakfast Spot     Mexican Restaurant   
2     Basketball Stadium           Hardware Store    Monument / Landmark   
3                   Café           Breakfast Spot         Sandwich Place   
4             Campground           Breakfast Spot         Sandwich Place   
5            Flower Shop        Other Repair Shop                   Park   
6              

How many venues do we have across these zip codes? 

In [36]:
venue_count = indy_venues.Neighborhood.isin(top_zip.Neighborhood)
indy_venues[venue_count].groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  Venue ID  \
Neighborhood                                                                   
46204                            50                      50     50        50   
46205                            50                      50     50        50   
46206                            50                      50     50        50   
46207                            50                      50     50        50   
46208                            47                      47     47        47   
46217                            42                      42     42        42   
46219                            50                      50     50        50   
46220                            50                      50     50        50   
46225                            50                      50     50        50   
46240                            50                      50     50        50   
46259                             8                       8      8         8   
46260                            32                      32     32        32   
46278                            35                      35     35        35   
46280                            45                      45     45        45   
46290                            35                      35     35        35   

              Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                   
46204                     50               50              50  
46205                     50               50              50  
46206                     50               50              50  
46207                     50               50              50  
46208                     47               47              47  
46217                     42               42              42  
46219                     50               50              50  
46220                     50               50              50  
46225                     50               50              50  
46240                     50               50              50  
46259                      8                8               8  
46260                     32               32              32  
46278                     35               35              35  
46280                     45               45              45  
46290                     35               35              35

We can only make 500 premium calls to get venue hours, so we'll need to choose less than 500 venues--in other words, only 10 of the 15 venues. Let's choose the 10 zips that seem to have the least amount of competition with the most potential attracting items. 

In [37]:
excluded_zip = [46220, 46240, 46260, 46290, 46219]
top_venues = top_zip[~top_zip['Neighborhood'].isin(excluded_zip)].reset_index(drop=True)
top_venues

Zip   Latitude  Longitude  Cluster Labels  Neighborhood  \
0  46259  39.651145 -85.980730               2         46259   
1  46204  39.771743 -86.155980               2         46204   
2  46278  39.897920 -86.286190               2         46278   
3  46217  39.668795 -86.183300               2         46217   
4  46280  39.939102 -86.138310               2         46280   
5  46205  39.824858 -86.138170               2         46205   
6  46208  39.820708 -86.171300               2         46208   
7  46207  39.767293 -86.160616               2         46207   
8  46225  39.746993 -86.159030               2         46225   
9  46206  39.761293 -86.161336               2         46206   

        1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Construction & Landscaping                 Diner                 Trail   
1          Mexican Restaurant                 Plaza           Coffee Shop   
2                       Hotel   American Restaurant            Food Truck   
3            Business Service   American Restaurant        Baseball Field   
4                      Bakery   American Restaurant   Rental Car Location   
5         American Restaurant   Rental Car Location        Science Museum   
6               Historic Site           Golf Course           Planetarium   
7                       Hotel                 Plaza        History Museum   
8          Mexican Restaurant           Art Gallery                 Hotel   
9                       Hotel    Mexican Restaurant   American Restaurant   

  4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
0                  Park      Business Service      Other Repair Shop   
1                  Café        Breakfast Spot    American Restaurant   
2        Shopping Plaza            Donut Shop                   Park   
3          Home Service    Mexican Restaurant                Dog Run   
4        Breakfast Spot                 Hotel            Golf Course   
5        Sandwich Place    Athletics & Sports                   Café   
6                 Trail                  Park                  River   
7    Mexican Restaurant                  Café         Breakfast Spot   
8        Sandwich Place        Breakfast Spot    American Restaurant   
9        History Museum              Boutique  General Entertainment   

  7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0     Food & Drink Shop               Dog Run             Donut Shop   
1    Basketball Stadium        Hardware Store    Monument / Landmark   
2                  Café        Breakfast Spot         Sandwich Place   
3            Campground        Breakfast Spot         Sandwich Place   
4           Flower Shop     Other Repair Shop                   Park   
5                  Fair                  Farm  General Entertainment   
6        Science Museum      Sculpture Garden                 Museum   
7   American Restaurant    Basketball Stadium  General Entertainment   
8    Basketball Stadium         Deli / Bodega           Concert Hall   
9      Football Stadium                  Park  Performing Arts Venue   

  10th Most Common Venue  
0            Fabric Shop  
1                   Park  
2       Sculpture Garden  
3           Skating Rink  
4          Deli / Bodega  
5          Historic Site  
6                 Office  
7       Football Stadium  
8            Coffee Shop  
9          Deli / Bodega

In [38]:
venue_count = indy_venues.Neighborhood.isin(top_venues.Neighborhood)
top_venues_id = indy_venues[venue_count].reset_index(drop=True)
top_venues_id

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0           46259              39.651145              -85.980730   
1           46259              39.651145              -85.980730   
2           46259              39.651145              -85.980730   
3           46259              39.651145              -85.980730   
4           46259              39.651145              -85.980730   
..            ...                    ...                     ...   
422         46206              39.761293              -86.161336   
423         46206              39.761293              -86.161336   
424         46206              39.761293              -86.161336   
425         46206              39.761293              -86.161336   
426         46206              39.761293              -86.161336   

                                                 Venue  \
0                                         HAWK DESIGNS   
1                                      Retherford Park   
2                                   Old Mcdonalds Cafe   
3                                   Dinner Bell Market   
4                            R V Medic Mobile Services   
..                                                 ...   
422                         Indiana Historical Society   
423  Eiteljorg Museum of American Indians & Western...   
424                                 Whole Foods Market   
425                             Historic Military Park   
426                               Indiana State Museum   

                     Venue ID  Venue Latitude  Venue Longitude  \
0    5898c8bbf595726e51a6eedd       39.649541       -85.978041   
1    4f61eac2e4b06b1a18061b43       39.651326       -85.971793   
2    4b9d42fdf964a5200c9e36e3       39.655417       -85.968074   
3    4b8078e9f964a5208f7530e3       39.655701       -85.968208   
4    594cf01be179107ef9a8ad05       39.637989       -85.970520   
..                        ...             ...              ...   
422  50df1b36e4b00e616a95b34c       39.770451       -86.165588   
423  4a972a5bf964a520c02820e3       39.768494       -86.167778   
424  5ab0279fa22db75fb69ef2bb       39.768426       -86.152236   
425  4bc9a1d168f976b0c0945d83       39.770730       -86.168125   
426  4ae08848f964a5200c8021e3       39.769061       -86.169334   

        Venue Category  
0     Business Service  
1                 Park  
2                Diner  
3        Grocery Store  
4    Other Repair Shop  
..                 ...  
422     History Museum  
423         Art Museum  
424      Grocery Store  
425               Park  
426     History Museum  

[427 rows x 8 columns]

## Venue Hours
This gives us the general call and lists the venues nearby, but we'll need to use each venue's ID to access popular hours.

In [40]:
venueHours_test = top_venues_id.loc[0,'Venue ID'] 
url_premium = 'https://api.foursquare.com/v2/venues/{}/hours?&client_id={}&client_secret={}&v={}'.format(
    venueHours_test,
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION)
print(url_premium)
hours_results_empty = requests.get(url_premium).json()['response']['popular']

venueHours_test = top_venues_id.loc[422,'Venue ID'] 
url_premium = 'https://api.foursquare.com/v2/venues/{}/hours?&client_id={}&client_secret={}&v={}'.format(
    venueHours_test,
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION)
print(url_premium)
hours_results = requests.get(url_premium).json()['response']['popular']

https://api.foursquare.com/v2/venues/5898c8bbf595726e51a6eedd/hours?&client_id=2DDTSCAMVOKHLPZTYIJSVR4TIVEXPIWIW0DM141PB5AXLMCP&client_secret=1GYDYKTOJQS33PND3MEVRBV2N1FMNDLDDG13TB3BMJK3HW3M&v=20180605
https://api.foursquare.com/v2/venues/50df1b36e4b00e616a95b34c/hours?&client_id=2DDTSCAMVOKHLPZTYIJSVR4TIVEXPIWIW0DM141PB5AXLMCP&client_secret=1GYDYKTOJQS33PND3MEVRBV2N1FMNDLDDG13TB3BMJK3HW3M&v=20180605


In [41]:
display(len(hours_results))
display(len(hours_results_empty))
# if there are no hours, ['timeframes'] will throw an error--add in later
venue_hours_raw = hours_results
venue_hours_raw

1

0

{'timeframes': [{'days': [4],
   'includesToday': True,
   'open': [{'start': '0700', 'end': '0900'},
    {'start': '1100', 'end': '2000'}],
   'segments': []},
  {'days': [5], 'open': [{'start': '0700', 'end': '2100'}], 'segments': []},
  {'days': [6], 'open': [{'start': '0800', 'end': '2200'}], 'segments': []},
  {'days': [7],
   'open': [{'start': '1000', 'end': '1100'},
    {'start': '1600', 'end': '1800'}],
   'segments': []},
  {'days': [1],
   'open': [{'start': '0700', 'end': '1000'},
    {'start': '1800', 'end': '1900'}],
   'segments': []},
  {'days': [2],
   'open': [{'start': '0700', 'end': '0800'},
    {'start': '1300', 'end': '1400'},
    {'start': '1700', 'end': '2000'}],
   'segments': []},
  {'days': [3], 'open': [{'start': '0700', 'end': '1900'}], 'segments': []}]}

Ideally, we would normalize to the 'timeframes'; however, one of the things I learned while going through a few URLs is that if a venue doesn't have hours available, it will throw an error if you normalize past the 'popular.' For this reason, we'll explore timeframes in another step, and later check to make sure hours are available before taking the step. 

In [42]:
display(json_normalize(venue_hours_raw['timeframes']))
display(json_normalize(venue_hours_raw['timeframes'][0])) #index of individual days 
display(json_normalize(venue_hours_raw['timeframes'][0])['open']) #['open'] has to be outside json_normalize or it throws an error

days includesToday                                               open  \
0  [4]          True  [{'start': '0700', 'end': '0900'}, {'start': '...   
1  [5]           NaN                 [{'start': '0700', 'end': '2100'}]   
2  [6]           NaN                 [{'start': '0800', 'end': '2200'}]   
3  [7]           NaN  [{'start': '1000', 'end': '1100'}, {'start': '...   
4  [1]           NaN  [{'start': '0700', 'end': '1000'}, {'start': '...   
5  [2]           NaN  [{'start': '0700', 'end': '0800'}, {'start': '...   
6  [3]           NaN                 [{'start': '0700', 'end': '1900'}]   

  segments  
0       []  
1       []  
2       []  
3       []  
4       []  
5       []  
6       []

days  includesToday                                               open  \
0  [4]           True  [{'start': '0700', 'end': '0900'}, {'start': '...   

  segments  
0       []

0    [{'start': '0700', 'end': '0900'}, {'start': '...
Name: open, dtype: object

In order to access the hours, we're going to need to extract the hours from the days. Each day could have multiple timeframes of popular hours in a dict with type string--we're going to need to pull those out, create a range between the start and end times, and then put them into a np array. We're going to do this for each day of the week for each venue, and then average across days to get a single list for each venue that can be used for clustering. 
Some venues don't have popular hours--the 'timeframes' are empty. We will use a list of 0s for these locations. 

In [43]:
def hoursPerDay(day):
    """Take a day from venue_days and return 1x24 numpy array with popular hours
    Arguments:
    day -- in the format of json_normalize(venue_hours_raw['timeframes'][index of day])['open']
    """
    hourList = np.array([0] * 24)
    for n in day:
#         print(n)
        start = int(n['start'][:-2])
        end = int(n['end'][:-2])
        timeRange = list(range(start-1,end)) # because of 0 indexing, time Range is one less than the actual hour
        hourList[timeRange] = 1
    return hourList

# example of how the function works
day_test = json_normalize(venue_hours_raw['timeframes'][1])['open'][0] # reminder: venue_days = json_normalize(venue_hours_raw['timeframes'][0])
hourList_test = np.array([0] * 24)
for n in day_test:
    print(n)
    start_test = int(n['start'][:-2])
    end_test = int(n['end'][:-2])
    timeRange_test = list(range(start_test-1,end_test)) # because of 0 indexing, time Range is one less than the actual hour
    hourList_test[timeRange_test] = 1
print(hourList_test)

{'start': '0700', 'end': '2100'}
[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0]


In [44]:
def hoursPerVenue(timeframe):
    """Take venue timeframes and return a single 1x24 np array of popular hours averaged across days
    Arguments:
    timeframe -- in the format of venue_hours_raw['timeframes'][0]
    """
    timeframe = json_normalize(timeframe)['open']
    hours_all = []
    for h in timeframe:
        hoursForDay = hoursPerDay(h)
        hours_all.append(hoursForDay)
    hours_all = np.stack(hours_all, axis=0)
    hours_mean = np.mean(hours_all, axis=0)
    return hours_mean


# example of how the function works
timeframe_test = venue_hours_raw['timeframes']
timeframe_test = json_normalize(timeframe_test)['open']
hours_all_test = []
for h in timeframe_test:
    print(h)
    hoursForDay_test = hoursPerDay(h)
    print(hoursForDay_test)
    hours_all_test.append(hoursForDay_test)
hours_all_test = np.stack(hours_all_test, axis=0)
hours_mean_test = np.mean(hours_all_test, axis=0)
print(hours_mean_test)
print('This list contains {} items for each hour of the day.'.format(len(hours_mean_test)))

[{'start': '0700', 'end': '0900'}, {'start': '1100', 'end': '2000'}]
[0 0 0 0 0 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0]
[{'start': '0700', 'end': '2100'}]
[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0]
[{'start': '0800', 'end': '2200'}]
[0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0]
[{'start': '1000', 'end': '1100'}, {'start': '1600', 'end': '1800'}]
[0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 1 1 1 0 0 0 0 0 0]
[{'start': '0700', 'end': '1000'}, {'start': '1800', 'end': '1900'}]
[0 0 0 0 0 0 1 1 1 1 0 0 0 0 0 0 0 1 1 0 0 0 0 0]
[{'start': '0700', 'end': '0800'}, {'start': '1300', 'end': '1400'}, {'start': '1700', 'end': '2000'}]
[0 0 0 0 0 0 1 1 0 0 0 0 1 1 0 0 1 1 1 1 0 0 0 0]
[{'start': '0700', 'end': '1900'}]
[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0]
[0.         0.         0.         0.         0.         0.
 0.71428571 0.85714286 0.71428571 0.71428571 0.71428571 0.57142857
 0.71428571 0.71428571 0.57142857 0.71428571 0.85714286 1.
 0.85714286 0.57142857 0.28571429 0.14285714 0.   

In [45]:
def getPopularHours(neighborhood, names, venueID):
    """Calls Foursquare to get popular hours for each venue and returns a dataframe containing hours
    Arguments:
    neighborhood -- neighborhood or zip code
    name -- names of venues
    venueID -- id of venues
    """
    hours_list = []
    for  neigh, name, venueid in zip(neighborhood, names, venueID):
        print(name)
        
        url_premium = 'https://api.foursquare.com/v2/venues/{}/hours?&client_id={}&client_secret={}&v={}'.format(
            venueid,
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION)
        
        hours_results = requests.get(url_premium).json()['response']['popular']

        if len(hours_results) == 0:
            hours = np.array([0]*24)
        else:
            hours = hoursPerVenue(timeframe = hours_results['timeframes'])
        
        hours = pd.Series(hours)
        venue_info = [neigh, name, venueid]
        venue_info.extend(hours)
        
        hours_list.append(venue_info)
    
    venue_popular = pd.DataFrame(hours_list)
    venue_popular.columns = ['Neighborhood', 'Venue', 'VenueID',
                       '01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12',
                       '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24',]
    
    return venue_popular

In [46]:
# test that the functions are working as intended
func_test_group = top_venues_id[10:15]
func_test_group
prem_test = getPopularHours(neighborhood = func_test_group['Neighborhood'], 
                            names = func_test_group['Venue'], 
                            venueID = func_test_group['Venue ID'])
prem_test

Bazbeaux Pizza
The Tap
Indiana World War Memorial
Hilbert Circle Theatre
Monument Circle


Neighborhood                       Venue                   VenueID   01  \
0         46204              Bazbeaux Pizza  4b1454fbf964a52059a123e3  0.0   
1         46204                     The Tap  55fdc727498e62f1d9ab9633  0.0   
2         46204  Indiana World War Memorial  4b144d59f964a520d4a023e3  0.0   
3         46204      Hilbert Circle Theatre  4b491f65f964a520dc6626e3  0.0   
4         46204             Monument Circle  4e78c67bfa769112d57ff55f  0.0   

    02   03   04   05        06        07  ...        15        16        17  \
0  0.0  0.0  0.0  0.0  0.000000  0.000000  ...  0.333333  0.333333  1.000000   
1  0.0  0.0  0.0  0.0  0.000000  0.000000  ...  0.166667  0.166667  0.666667   
2  0.0  0.0  0.0  0.0  0.142857  0.142857  ...  0.714286  1.000000  1.000000   
3  0.0  0.0  0.0  0.0  0.000000  0.000000  ...  0.285714  0.714286  0.714286   
4  0.0  0.0  0.0  0.0  0.000000  0.000000  ...  0.000000  0.000000  0.000000   

         18        19        20        21        22        23   24  
0  1.000000  1.000000  1.000000  1.000000  0.500000  0.333333  0.0  
1  0.666667  0.666667  0.666667  0.666667  0.500000  0.166667  0.0  
2  0.714286  0.571429  0.428571  0.285714  0.285714  0.000000  0.0  
3  0.714286  0.857143  0.714286  0.428571  0.285714  0.000000  0.0  
4  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.0  

[5 rows x 27 columns]

In [47]:
# Get venue hours for all venues
venue_popular_hrs = getPopularHours(neighborhood = top_venues_id['Neighborhood'], 
                            names = top_venues_id['Venue'], 
                            venueID = top_venues_id['Venue ID'])

HAWK DESIGNS
Retherford Park
Old Mcdonalds Cafe
Dinner Bell Market
R V Medic Mobile Services
JT Drywall
Harmony Walking Trail
Mainsource Supply LLC
The Eagle
Bakersfield Mass Ave
Bazbeaux Pizza
The Tap
Indiana World War Memorial
Hilbert Circle Theatre
Monument Circle
Cultural Trail - Downtown Indianapolis
Rocket Fizz
Tomlinson Tap Room
Bru Burger Bar
Silver in the City
Indy Bike Hub YMCA
Potbelly Sandwich Shop
American Legion Mall
Rebar Indy
Whole Foods Market
MacNiven's Restaurant & Bar
Coat Check Coffee
Fogo de Chao
The Rathskeller
Fat Dan's Deli
Goodfellas Pizzeria
Wild Eggs
Indiana Repertory Theatre
Fusek's True Value
Canal Walk
Athenaeum
Rathskeller Biergarten
Sub Zero Nitrogen Ice Cream
Chipotle Mexican Grill
St. Elmo Steak House
Red's Classic Barber Shop Co.
Libertine on Mass
Kroger
Cafe Patachou
Chatterbox Jazz Club
Yard House
Harry & Izzy's
Indiana Historical Society
Bankers Life Fieldhouse
Nine Irish Brothers
Napolese Artisanal Pizzeria
YMCA @ the Athenaeum
Sun King Brewery
S

retro101
Indy Bike Hub YMCA
Iozzo's Garden Of Italy
Indiana Historical Society
Eiteljorg Museum of American Indians & Western Art
Whole Foods Market
Historic Military Park
Indiana State Museum


In [49]:
display(venue_popular_hrs.head())
venue_popular_hrs.tail()

Neighborhood                      Venue                   VenueID   01  \
0         46259               HAWK DESIGNS  5898c8bbf595726e51a6eedd  0.0   
1         46259            Retherford Park  4f61eac2e4b06b1a18061b43  0.0   
2         46259         Old Mcdonalds Cafe  4b9d42fdf964a5200c9e36e3  0.0   
3         46259         Dinner Bell Market  4b8078e9f964a5208f7530e3  0.0   
4         46259  R V Medic Mobile Services  594cf01be179107ef9a8ad05  0.0   

    02   03   04   05   06   07  ...   15   16   17   18   19   20   21   22  \
0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    23   24  
0  0.0  0.0  
1  0.0  0.0  
2  0.0  0.0  
3  0.0  0.0  
4  0.0  0.0  

[5 rows x 27 columns]

Neighborhood                                              Venue  \
422         46206                         Indiana Historical Society   
423         46206  Eiteljorg Museum of American Indians & Western...   
424         46206                                 Whole Foods Market   
425         46206                             Historic Military Park   
426         46206                               Indiana State Museum   

                      VenueID   01   02   03   04   05        06        07  \
422  50df1b36e4b00e616a95b34c  0.0  0.0  0.0  0.0  0.0  0.000000  0.714286   
423  4a972a5bf964a520c02820e3  0.0  0.0  0.0  0.0  0.0  0.000000  0.000000   
424  5ab0279fa22db75fb69ef2bb  0.0  0.0  0.0  0.0  0.0  0.000000  0.000000   
425  4bc9a1d168f976b0c0945d83  0.0  0.0  0.0  0.0  0.0  0.166667  0.166667   
426  4ae08848f964a5200c8021e3  0.0  0.0  0.0  0.0  0.0  0.000000  0.285714   

     ...        15        16        17        18        19        20  \
422  ...  0.571429  0.714286  0.857143  1.000000  0.857143  0.571429   
423  ...  1.000000  0.857143  0.571429  0.714286  0.571429  0.428571   
424  ...  0.857143  0.857143  1.000000  1.000000  1.000000  1.000000   
425  ...  0.500000  0.500000  0.500000  0.333333  0.500000  0.500000   
426  ...  1.000000  0.857143  0.428571  0.571429  0.571429  0.428571   

           21        22        23   24  
422  0.285714  0.142857  0.000000  0.0  
423  0.142857  0.142857  0.142857  0.0  
424  0.571429  0.000000  0.000000  0.0  
425  0.333333  0.166667  0.000000  0.0  
426  0.142857  0.142857  0.000000  0.0  

[5 rows x 27 columns]

### Cluster using the Venue hours
The benefit to breaking it out by hour as the functions above have done is that we should be able to use those directly for clustering. Before we do, let's take a look at some more basic exploration. What is the mean of hours across ZIP codes? Will clustering agree with this spread?

In [50]:
venue_popular_hrs.groupby('Neighborhood').mean()

01        02        03        04        05        06  \
Neighborhood                                                               
46204         0.000000  0.000000  0.000000  0.000000  0.028571  0.058095   
46205         0.000000  0.002857  0.002857  0.000000  0.014286  0.017619   
46206         0.002857  0.002857  0.000000  0.000000  0.014286  0.034762   
46207         0.000000  0.000000  0.000000  0.000000  0.025714  0.063810   
46208         0.003040  0.003040  0.000000  0.000000  0.000000  0.003546   
46217         0.000000  0.003401  0.003401  0.000000  0.017007  0.020408   
46225         0.003333  0.002857  0.002857  0.000000  0.005714  0.005714   
46259         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
46278         0.000000  0.000000  0.000000  0.008163  0.028571  0.104762   
46280         0.003175  0.003175  0.000000  0.000000  0.043386  0.055820   

                    07        08        09        10  ...        15        16  \
Neighborhood                                          ...                       
46204         0.157143  0.199619  0.223810  0.226476  ...  0.435619  0.468286   
46205         0.060762  0.107524  0.149143  0.194857  ...  0.306667  0.335524   
46206         0.076762  0.110667  0.146381  0.180000  ...  0.346476  0.393714   
46207         0.152381  0.183238  0.211524  0.225143  ...  0.426571  0.434762   
46208         0.066768  0.156231  0.181155  0.217629  ...  0.379129  0.374873   
46217         0.072562  0.106576  0.160998  0.168367  ...  0.217120  0.224490   
46225         0.039524  0.073333  0.116667  0.142857  ...  0.253714  0.277048   
46259         0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   
46278         0.186259  0.252789  0.268435  0.311973  ...  0.379592  0.323810   
46280         0.119947  0.260847  0.326984  0.342540  ...  0.368254  0.453439   

                    17        18        19        20        21        22  \
Neighborhood                                                               
46204         0.696476  0.712571  0.667524  0.572381  0.435333  0.262762   
46205         0.353238  0.363524  0.340000  0.321619  0.250190  0.116095   
46206         0.557143  0.618571  0.615238  0.559048  0.444571  0.254905   
46207         0.616095  0.650190  0.608000  0.546667  0.427714  0.226381   
46208         0.373658  0.355015  0.319656  0.303141  0.192199  0.089666   
46217         0.258503  0.258503  0.231293  0.208617  0.193311  0.113946   
46225         0.571429  0.634286  0.647619  0.636190  0.536000  0.331333   
46259         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
46278         0.426531  0.435374  0.390476  0.307483  0.216327  0.072109   
46280         0.517989  0.541799  0.521693  0.424868  0.277354  0.084656   

                    23        24  
Neighborhood                      
46204         0.093524  0.000000  
46205         0.032381  0.000000  
46206         0.091857  0.002857  
46207         0.071905  0.000000  
46208         0.015704  0.003040  
46217         0.027211  0.000000  
46225         0.130190  0.003333  
46259         0.000000  0.000000  
46278         0.024490  0.000000  
46280         0.015873  0.003175  

[10 rows x 24 columns]

We can see 46259 has a mean of 0.0 across all hours, but looking at the number of venues per zip from a previous cell shows that this zip code has less than 10 venues, which likely contributes to this. We can remove this zip from the clustering to ensure it doesn't skew the results. What would be the zip codes we might predict would be clustered for morning hours? 

In [61]:
venue_popular_hrs = venue_popular_hrs[venue_popular_hrs.Neighborhood != 46259]

419


419

In [66]:
display(venue_popular_hrs.loc[:,'Neighborhood':'12'].groupby('Neighborhood').mean())
venue_popular_hrs.loc[:,'Neighborhood':'12'].groupby('Neighborhood').mean().mean(axis=1).sort_values()

01        02        03        04        05        06  \
Neighborhood                                                               
46204         0.000000  0.000000  0.000000  0.000000  0.028571  0.058095   
46205         0.000000  0.002857  0.002857  0.000000  0.014286  0.017619   
46206         0.002857  0.002857  0.000000  0.000000  0.014286  0.034762   
46207         0.000000  0.000000  0.000000  0.000000  0.025714  0.063810   
46208         0.003040  0.003040  0.000000  0.000000  0.000000  0.003546   
46217         0.000000  0.003401  0.003401  0.000000  0.017007  0.020408   
46225         0.003333  0.002857  0.002857  0.000000  0.005714  0.005714   
46278         0.000000  0.000000  0.000000  0.008163  0.028571  0.104762   
46280         0.003175  0.003175  0.000000  0.000000  0.043386  0.055820   

                    07        08        09        10        11        12  
Neighborhood                                                              
46204         0.157143  0.199619  0.223810  0.226476  0.440952  0.606857  
46205         0.060762  0.107524  0.149143  0.194857  0.287429  0.342190  
46206         0.076762  0.110667  0.146381  0.180000  0.403571  0.536333  
46207         0.152381  0.183238  0.211524  0.225143  0.438667  0.604000  
46208         0.066768  0.156231  0.181155  0.217629  0.329787  0.386018  
46217         0.072562  0.106576  0.160998  0.168367  0.206916  0.241497  
46225         0.039524  0.073333  0.116667  0.142857  0.364952  0.488667  
46278         0.186259  0.252789  0.268435  0.311973  0.495646  0.552789  
46280         0.119947  0.260847  0.326984  0.342540  0.557460  0.593651

Neighborhood
46217    0.083428
46205    0.098294
46225    0.103873
46208    0.112268
46206    0.125706
46207    0.158706
46204    0.161794
46278    0.184116
46280    0.192249
dtype: float64

From a quick glance, it looks like 46280, 46278, 46204, and 46207 may be some of our likely contenders. Let's cluster to compare against a data-driven model that may find trends we don't easily see here.

We're going to use less clusters for this--3, hoping to break it down by morning, afternoon, and night. 

In [69]:
# set number of clusters
kclusters = 3

venue_popular_clustering = venue_popular_hrs.drop(['Neighborhood', 'Venue', 'VenueID'], 1).reset_index(drop=True)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(venue_popular_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 2, 0, 1, 2, 0, 0])

In [70]:
venue_popular_hrs.insert(0, 'Cluster Labels', kmeans.labels_)
venue_popular_hrs.head()

Cluster Labels  Neighborhood                       Venue  \
8                0         46204                   The Eagle   
9                0         46204        Bakersfield Mass Ave   
10               0         46204              Bazbeaux Pizza   
11               0         46204                     The Tap   
12               2         46204  Indiana World War Memorial   

                     VenueID   01   02   03   04   05        06  ...  \
8   5671f9b5498ece84b31f180c  0.0  0.0  0.0  0.0  0.0  0.000000  ...   
9   512ffd44e4b0e80cc96198fb  0.0  0.0  0.0  0.0  0.0  0.000000  ...   
10  4b1454fbf964a52059a123e3  0.0  0.0  0.0  0.0  0.0  0.000000  ...   
11  55fdc727498e62f1d9ab9633  0.0  0.0  0.0  0.0  0.0  0.000000  ...   
12  4b144d59f964a520d4a023e3  0.0  0.0  0.0  0.0  0.0  0.142857  ...   

          15        16        17        18        19        20        21  \
8   0.142857  0.142857  0.714286  0.714286  0.714286  0.714286  0.714286   
9   0.142857  0.142857  0.714286  0.714286  0.714286  0.714286  0.714286   
10  0.333333  0.333333  1.000000  1.000000  1.000000  1.000000  1.000000   
11  0.166667  0.166667  0.666667  0.666667  0.666667  0.666667  0.666667   
12  0.714286  1.000000  1.000000  0.714286  0.571429  0.428571  0.285714   

          22        23   24  
8   0.714286  0.000000  0.0  
9   0.428571  0.285714  0.0  
10  0.500000  0.333333  0.0  
11  0.500000  0.166667  0.0  
12  0.285714  0.000000  0.0  

[5 rows x 28 columns]

So now that we have the venues clustered, we can start taking a look at which hours are most common to which clusters. 

In [110]:
venue_hrs_by_cluster = venue_popular_hrs.groupby('Cluster Labels').mean()
display(venue_hrs_by_cluster)
print('Cluster 0')
display(venue_hrs_by_cluster.loc[0].sort_values(ascending = False)[1:15])
print('Cluster 1')
display(venue_hrs_by_cluster.loc[1].sort_values(ascending = False)[1:15])
print('Cluster 2')
display(venue_hrs_by_cluster.loc[2].sort_values(ascending = False)[1:15])

Neighborhood        01        02        03        04  \
Cluster Labels                                                         
0               46220.984293  0.003116  0.002992  0.000748  0.000000   
1               46223.762963  0.000000  0.001058  0.001058  0.000000   
2               46227.462366  0.000000  0.001536  0.001536  0.003072   

                      05        06        07        08        09  ...  \
Cluster Labels                                                    ...   
0               0.008227  0.024308  0.029045  0.047295  0.093343  ...   
1               0.000000  0.002469  0.006914  0.005926  0.009947  ...   
2               0.070148  0.120712  0.386559  0.605274  0.673016  ...   

                      15        16        17        18        19        20  \
Cluster Labels                                                               
0               0.453403  0.504488  0.782299  0.859686  0.870406  0.826901   
1               0.005714  0.010794  0.039577  0.040423  0.045644  0.052346   
2               0.622325  0.606861  0.554992  0.499590  0.359037  0.211623   

                      21        22        23        24  
Cluster Labels                                          
0               0.663301  0.338731  0.100960  0.003116  
1               0.041764  0.041764  0.029771  0.000000  
2               0.102867  0.049155  0.012289  0.000000  

[3 rows x 25 columns]

Cluster 0


19    0.870406
18    0.859686
20    0.826901
17    0.782299
13    0.680155
21    0.663301
12    0.636213
14    0.596011
16    0.504488
15    0.453403
11    0.437659
22    0.338731
10    0.125804
23    0.100960
Name: 0, dtype: float64

Cluster 1


20    0.052346
19    0.045644
21    0.041764
22    0.041764
18    0.040423
17    0.039577
23    0.029771
16    0.010794
09    0.009947
10    0.009947
13    0.008889
12    0.008677
14    0.007831
07    0.006914
Name: 1, dtype: float64

Cluster 2


12    0.866257
13    0.861137
11    0.848643
14    0.761623
10    0.717512
09    0.673016
15    0.622325
16    0.606861
08    0.605274
17    0.554992
18    0.499590
07    0.386559
19    0.359037
20    0.211623
Name: 2, dtype: float64

It looks like clusters 0 and 1 both have an emphasis on afternoon, evening, or night hours. While cluster 2 does have a few later hours, it seems more oriented towards morning or midday hours than the other two, so this will be our focus. But which zip codes have the highest number of venues in cluster 2?

In [129]:
# get number of cluster labels per neighborhood, ranked from most to least
clusters_ranked = venue_popular_hrs.groupby('Neighborhood')['Cluster Labels'].value_counts()
display(clusters_ranked)

Neighborhood  Cluster Labels
46204         0                 29
              2                 14
              1                  7
46205         1                 25
              0                 16
              2                  9
46206         0                 30
              1                 10
              2                 10
46207         0                 29
              2                 13
              1                  8
46208         1                 22
              0                 15
              2                 10
46217         1                 27
              0                  8
              2                  7
46225         0                 35
              1                 10
              2                  5
46278         1                 13
              0                 11
              2                 11
46280         0                 18
              2                 14
              1                 13
Name: Cluster Labels, dtyp

It looks like none of the zip codes have cluster 2 as their highest ranked number of venues, so let's just look at which has the highest number overall.

In [145]:
display(clusters_ranked[:,2])
# let's get the ratio per number of overall cluster labels
(clusters_ranked[:,2])/(venue_popular_hrs.groupby('Neighborhood')['Cluster Labels'].sum())

Neighborhood
46204    14
46205     9
46206    10
46207    13
46208    10
46217     7
46225     5
46278    11
46280    14
Name: Cluster Labels, dtype: int64

Neighborhood
46204    0.400000
46205    0.209302
46206    0.333333
46207    0.382353
46208    0.238095
46217    0.170732
46225    0.250000
46278    0.314286
46280    0.341463
Name: Cluster Labels, dtype: float64

## Final verdict
Based on the results above, it seems that zip code 46204 or 46207 would be the best bet for opening a breakfast spot that has relevant nearby venues without too much competition and popular hours skewing towards the earlier hours of the day.